include/nbl/builtin/hlsl/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,596 @@
+           }
+           else
+           {
+              frac.xyz = shift64ExtraRightJamming(frac.x, frac.y, 0u, -shiftCount);
+           }
+           exp -= shiftCount;
+           return roundAndPackFloat64(sign, exp, frac.x, frac.y, frac.z);
+        }
+
+    }
+
+    struct emulated_float64_t
+    {
+        using storage_t = uint64_t;
+
+        storage_t data;
+
+        // constructors
+        // TODO: specializations
+        template <typename T>
+        static emulated_float64_t create(T val)
+        {
+#ifndef __HLSL_VERSION
+            emulated_float64_t output;
+            output.data = reinterpret_cast<storage_t&>(val);
+            return output;
+#else
+            uint32_t lowBits;
+            uint32_t highBits;
+            asuint(val, lowBits, highBits);
+            
+            emulated_float64_t output;
+            output.data = (uint64_t(highBits) << 32) | uint64_t(lowBits);
+            return output;
+#endif
+        }
+        
+        static emulated_float64_t createEmulatedFloat64PreserveBitPattern(uint64_t val)
+        {
+            emulated_float64_t output;
+            output.data = val;
+            return output;
+        }
+        
+        // TODO: won't not work for uints with msb of index > 52
+#ifndef __HLSL_VERSION
+        template<>
+        static emulated_float64_t create(uint64_t val)
+        {
+#ifndef __HLSL_VERSION
+            const uint64_t msbIndex = nbl::hlsl::findMSB(val);
+#else
+            const uint64_t msbIndex = firstbithigh(val);
+#endif
+            uint64_t exp = ((msbIndex + 1023) << 52) & 0x7FF0000000000000;
+            uint64_t mantissa = (val << (52 - msbIndex)) & 0x000FFFFFFFFFFFFFull;
+            emulated_float64_t output;
+            output.data = exp | mantissa;
+            return output;
+        }
+#endif        
+        // TODO: temporary, remove
+#ifndef __HLSL_VERSION
+        template<>
+        static emulated_float64_t create(double val)
+        {
+            emulated_float64_t output;
+            output.data = reinterpret_cast<uint64_t&>(val);
+            return output;
+        }
+#endif
+
+        // arithmetic operators
+        emulated_float64_t operator+(const emulated_float64_t rhs)
+        {
+            emulated_float64_t retval = createEmulatedFloat64PreserveBitPattern(0u);
+            
+            uint32_t lhsSign = uint32_t((data & 0x8000000000000000ull) >> 32);
+            uint32_t rhsSign = uint32_t((rhs.data & 0x8000000000000000ull) >> 32);
+
+            uint32_t lhsLow = uint32_t(data & 0x00000000FFFFFFFFull);
+            uint32_t rhsLow = uint32_t(rhs.data & 0x00000000FFFFFFFFull);
+            uint32_t lhsHigh = uint32_t((data & 0x000FFFFF00000000ull) >> 32);
+            uint32_t rhsHigh = uint32_t((rhs.data & 0x000FFFFF00000000ull) >> 32);
+
+            int lhsExp = int((data >> 52) & 0x7FFull);
+            int rhsExp = int((rhs.data >> 52) & 0x7FFull);

```

Comment:
I'd call them `biasedExp` not just `Exp` cause they have that implicit `+1023` inside themselves